File: app/src/main/java/com/boardgamegeek/provider/CategoriesIdProvider.java
package com.boardgamegeek.provider;

import android.net.Uri;

import com.boardgamegeek.provider.BggContract.Categories;
import com.boardgamegeek.provider.BggDatabase.Tables;
import com.boardgamegeek.util.SelectionBuilder;

public class CategoriesIdProvider extends BaseProvider {

	@Override
	protected SelectionBuilder buildSimpleSelection(Uri uri) {
		int categoryId = Categories.getCategoryId(uri);
		return new SelectionBuilder().table(Tables.CATEGORIES).whereEquals(Categories.CATEGORY_ID, categoryId);
	}

	@Override
	protected String getPath() {
		return "categories/#";
	}

	@Override
	protected String getType(Uri uri) {
		return Categories.CONTENT_ITEM_TYPE;
	}
}

File: app/src/main/java/com/boardgamegeek/provider/CategoriesIdProvider.kt
package com.boardgamegeek.provider

import android.net.Uri
import android.provider.BaseColumns
import com.boardgamegeek.provider.BggContract.Categories
import com.boardgamegeek.provider.BggContract.Companion.PATH_CATEGORIES
import com.boardgamegeek.provider.BggDatabase.Tables

class CategoriesIdProvider : BaseProvider() {
    override fun getType(uri: Uri) = Categories.CONTENT_ITEM_TYPE

    override val path = "$PATH_CATEGORIES/#"

    private val table = Tables.CATEGORIES

    override fun buildSimpleSelection(uri: Uri): SelectionBuilder {
        val categoryId = Categories.getCategoryId(uri)
        return SelectionBuilder()
            .table(table)
            .whereEquals(Categories.Columns.CATEGORY_ID, categoryId)
    }

    override fun buildExpandedSelection(uri: Uri, projection: Array<String>?): SelectionBuilder {
        val categoryId = Categories.getCategoryId(uri)
        val builder = SelectionBuilder()
            .mapToTable(BaseColumns._ID, table)
            .mapToTable(Categories.Columns.CATEGORY_ID, table)
            .whereEquals(Categories.Columns.CATEGORY_ID, categoryId)
        if (projection.orEmpty().contains(Categories.Columns.ITEM_COUNT)) {
            builder
                .table(Tables.CATEGORIES_JOIN_COLLECTION)
                .groupBy("$table.${Categories.Columns.CATEGORY_ID}")
                .mapAsCount(Categories.Columns.ITEM_COUNT)
        } else {
            builder.table(table)
        }
        return builder
    }
}
